### PR TITLE
Feat/latency store

### PR DESCRIPTION
**Pull request by d-v-b**
This PR adds a new store class, `WrapperStore`, that implements `Store` but wraps an existing store instance and delegates all of its methods to those of the wrapped store. I  then use `WrapperStore` to define a store class that exists for performance testing, called `LatencyStore`, which (at the moment) adds parametric latency to the `set` and `get` methods.

TODO:
* [ ] Add unit tests and/or doctests in docstrings
* [ ] Add docstrings and API docs for any new/modified user-facing classes and functions
* [ ] New/modified features documented in docs/tutorial.rst
* [ ] Changes documented in docs/release.rst
* [ ] GitHub Actions have all passed
* [ ] Test coverage is 100% (Codecov passes)

### CODE REVIEW

**Comment by d-v-b** [comment]
anyone interested in implementing n5 in zarr-python 3 might find a use for `WrapperStore` (cc @dstansby )

File: src/zarr/storage/wrapper.py [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+from zarr.abc.store import Store
+
+T_Store = TypeVar("T_Store", bound=Store)
+
+
+class WrapperStore(Store, Generic[T_Store]):
+    """
+    A store class that wraps an existing ``Store`` instance.
+    By default all of the store methods are delegated to the wrapped store instance, which is
+    accessible via the ``._wrapped`` attribute of this class.
```
Comment (jhamman) [inline src/zarr/storage/wrapper.py:23]:
```suggestion
    accessible via the ``._store`` attribute of this class.
```

File: src/zarr/storage/logging.py [hunk 2]
```diff
@@ -203,19 +205,19 @@ async def set_partial_values(
         with self.log(keys):
             return await self._store.set_partial_values(key_start_values=key_start_values)
 
-    async def list(self) -> AsyncIterator[str]:
+    async def list(self) -> AsyncGenerator[str, None]:
```
Comment (jhamman) [inline src/zarr/storage/logging.py:208]: What's the reason for switching these back to `AsyncGenerators`? 

xref: https://github.com/zarr-developers/zarr-python/pull/2381

File: src/zarr/testing/store.py [hunk 3]
```diff
@@ -319,25 +330,43 @@ async def test_set_if_not_exists(self, store: S) -> None:
...
+        Parameters
+        ----------
+        key : str
+        prototype : BufferPrototype
+        byte_range : ByteRangeRequest, optional
+
+        Returns
+        -------
+        buffer : Buffer or None
+        """
```
Comment (jhamman) [inline src/zarr/testing/store.py:370]: either remove this to use the base classes doc string or add a similar one to `set`

**Review by jhamman — COMMENTED** [review]
Looks good. Need to sort out the AsyncIterator type question but otherwise this looks great.

File: src/zarr/storage/logging.py (hunk 2 above)
Comment (d-v-b) [reply src/zarr/storage/logging.py:208]: this was likely me blindly obeying a wayward linter. I will revert

File: src/zarr/testing/store.py (hunk 3 above)
Comment (d-v-b) [reply src/zarr/testing/store.py]: i went for the latter

**Comment by d-v-b** [comment]
@jhamman the type annotations and docstrings are ironed out

**Review by dcherian — APPROVED** [review]

**Pull request merged**